esphome/components/sen0501/sen0501.cpp

```diff
@@ -0,0 +1,166 @@
+// Device registers
+static const uint16_t REG_PID = 0x0000;
+static const uint16_t REG_VID = 0x0002;
+static const uint16_t REG_DEVICE_ADDR = 0x0004;
+static const uint16_t REG_UART_CTRL0 = 0x0006;
+static const uint16_t EG_UART_CTRL1 = 0x0008;
+static const uint16_t REG_VERSION = 0x000A;
+// Data registers
+static const uint16_t REG_ULTRAVIOLET_INTENSITY = 0x0010;
+static const uint16_t REG_LUMINOUS_INTENSITY = 0x0012;
+static const uint16_t REG_TEMP = 0x0014;
+static const uint16_t REG_HUMIDITY = 0x0016;
+static const uint16_t REG_ATMOSPHERIC_PRESSURE = 0x0018;
+
+// PUBLIC
+
+void Sen0501Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up sen0501...");
+  uint8_t product_id_first_byte;
+  if (!this->read_byte(REG_PID, &product_id_first_byte)) {
+    this->error_code_ = COMMUNICATION_FAILED;
+    this->mark_failed();
+    return;
+  } else {
+    uint8_t buf[2];
+    this->read_bytes(REG_PID, buf, 2);
+    uint16_t product_id = buf[0] << 8 | buf[1];
+    if ((product_id != DEVICE_PID_GRAVITY) && (product_id != DEVICE_PID_BREAKOUT)) {
+      this->error_code_ = WRONG_DEVICE_ID;
+      this->mark_failed();
+      return;
+    }
+    this->read_bytes(REG_VID, buf, 2);
+    uint16_t vendor_id = buf[0] << 8 | buf[1];
+    if (vendor_id != DEVICE_VID) {
+      this->error_code_ = WRONG_VENDOR_ID;
+      this->mark_failed();
+      return;
+    }
+  }
+}
+
+void Sen0501Component::update() {
+  this->read_temperature_();
+  this->read_humidity_();
+  this->read_uv_intensity_();
+  this->read_luminous_intensity_();
+  this->read_atmospheric_pressure_();
+}
+
+void Sen0501Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "DFRobot Environmental Sensor - sen0501:");
+  LOG_I2C_DEVICE(this);
+  switch (this->error_code_) {
+    case COMMUNICATION_FAILED:
+      ESP_LOGE(TAG, "Communication with sen0501 failed!");
+      break;
+    case WRONG_DEVICE_ID:
+      ESP_LOGE(TAG, "Wrong device ID! Is it a sen0501?");
+      break;
+    case WRONG_VENDOR_ID:
+      ESP_LOGE(TAG, "Wrong vendor ID! Is it made by DFRobot(Zhiwei Robotics Corp)?");
+      break;
+    case NONE:
+    default:
+      break;
+  }
+  LOG_UPDATE_INTERVAL(this);
+  LOG_SENSOR("  ", "Temperature", this->temperature_);
+  LOG_SENSOR("  ", "Humidity", this->humidity_);
+  LOG_SENSOR("  ", "UV Intensity", this->uv_intensity_);
+  LOG_SENSOR("  ", "Luminous Intensity", this->luminous_intensity_);
+  LOG_SENSOR("  ", "Atmospheric Pressure", this->atmospheric_pressure_);
+  LOG_SENSOR("  ", "Elevation", this->elevation_);
+}
+
+float Sen0501Component::get_setup_priority() const { return setup_priority::DATA; }
+
+// PROTECTED
+
+float Sen0501Component::map_float_(float x, float in_min, float in_max, float out_min, float out_max) {
+  return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
+}
+
+void Sen0501Component::read_temperature_() {
+  uint8_t buffer[2];
+  uint16_t data;
+  float temp;
+  read_bytes(REG_TEMP, buffer, 2);
+  data = buffer[0] << 8 | buffer[1];
```

Comment:
All member variables and function calls should be prefixed with `this->`
We also have a helper for bitshift adding bytes together with casting.
Please apply these below too

```suggestion
  float temp;
  this->read_bytes(REG_TEMP, buffer, 2);
  uint16_t data = encode_uint16(buffer[0], buffer[1]);
```